### PR TITLE
sound: add clippy lint checks 

### PR DESCRIPTION
**Pull request by epilys**
Depends on #513

### Summary of the PR

Add basic lints to deny unidiomatic or potentially bad code.

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

**Review by vireshk — APPROVED** [review]

**Comment by vireshk** [comment]
@stefano-garzarella @stsquad 

I managed to screw up and merge this without approval from you guys. I was trying to enable auto-merge only :(

Can you guys please check the pull request again and see if there is something worth fixing ?

File: staging/vhost-device-sound/Cargo.toml [hunk 1]
```diff
@@ -6,6 +6,7 @@ description = "A virtio-sound device using the vhost-user protocol."
 repository = "https://github.com/rust-vmm/vhost-device"
 readme = "README.md"
 keywords = ["vhost", "sound", "virtio-sound", "virtio-snd", "virtio"]
+categories = ["multimedia::audio"]
```
Comment (stefano-garzarella) [inline staging/vhost-device-sound/Cargo.toml:9]: What about adding also `virtualization`?

File: staging/vhost-device-sound/src/lib.rs [hunk 2]
```diff
@@ -1,5 +1,34 @@
...
+    clippy::borrow_as_ptr,
+    clippy::case_sensitive_file_extension_comparisons,
+    clippy::cast_lossless,
+    clippy::cast_ptr_alignment,
+    clippy::naive_bytecount
+)]
+#![allow(
+    clippy::significant_drop_in_scrutinee,
+    clippy::significant_drop_tightening
+)]
```
Comment (stefano-garzarella) [inline staging/vhost-device-sound/src/lib.rs:31]: Is it possible to enable these in some way for the entire workspace?

Comment (stefano-garzarella) [reply staging/vhost-device-sound/src/lib.rs:31]: Out of curiosity, on what basis did we decide which ones to enable?

**Comment by stefano-garzarella** [comment]
> @stefano-garzarella @stsquad
> 
> I managed to screw up and merge this without approval from you guys. I was trying to enable auto-merge only :(

No problem, can happen ;-)

> 
> Can you guys please check the pull request again and see if there is something worth fixing ?

I left some comments that we can surely fix in another PR.
IIUC this also includes #513, so should we close it?

**Comment by vireshk** [comment]
Done

File: staging/vhost-device-sound/Cargo.toml (hunk 1 above)
Comment (epilys) [reply staging/vhost-device-sound/Cargo.toml:9]: I missed the most obvious one in the category slug list :/ thanks! 😃

File: staging/vhost-device-sound/src/lib.rs (hunk 2 above)
Comment (epilys) [reply staging/vhost-device-sound/src/lib.rs:31]: I took the basic categories from here:

https://doc.rust-lang.org/nightly/clippy/

```text
clippy::correctness	code that is outright wrong or useless	deny
clippy::suspicious	code that is most likely wrong or useless	warn
clippy::complexity	code that does something simple but in a complex way	warn
clippy::perf	code that can be written to run faster	warn
clippy::style	code that should be written in a more idiomatic way	warn
clippy::pedantic	lints which are rather strict or might have false positives	allow
clippy::nursery	new lints that are still under development	allow
clippy::cargo	lints for the cargo manifest	allow
```

And enabled all of them, then disabled `pedantic` and only cherry picked some of them.

For context, pedantic has some... really pedantic lints like:

```text
error: item name starts with its containing module's name
   --> vhost-device-sound/src/virtio_sound.rs:269:12
    |
269 | pub struct VirtioSoundPcmHeader {
    |            ^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#module_name_repetitions
```

Also no wildcard imports, etc.

Comment (epilys) [reply staging/vhost-device-sound/src/lib.rs:31]: > Is it possible to enable these in some way for the entire workspace?

Good question. It's gonna be possible in stable soon:

https://rust-lang.github.io/rfcs/3389-manifest-lint.html

Quoting the RFC:

```text
Currently, you can configure lints through

    #[<level>(<lint>)] or #![<level>(<lint>)], like #[forbid(unsafe)]
        But this doesn't scale up with additional targets (benches, examples, tests) or workspaces
    On the command line, like cargo clippy -- --forbid unsafe
        This puts the burden on the caller
    Through RUSTFLAGS, like RUSTFLAGS=--forbid=unsafe cargo clippy
        This puts the burden on the caller
    In .cargo/config.toml's target.*.rustflags
        This couples you to the running in specific directories and not running in the right directory causes rebuilds
        The cargo team has previously stated that [they would like to see package-specific config moved to manifests](https://internals.rust-lang.org/t/proposal-move-some-cargo-config-settings-to-cargo-toml/13336/14?u=epage)
```

So to enable them for the workspace we could put them in RUSTFLAGS for all targets in a `.cargo/config.toml`.

Comment (stefano-garzarella) [reply staging/vhost-device-sound/src/lib.rs:31]: > I took the basic categories from here:
> 
> https://doc.rust-lang.org/nightly/clippy/
> 

Cool, thanks for sharing!
I'm thinking about whether we should add a comment here with this link, but maybe not.

> 
> And enabled all of them, then disabled `pedantic` and only cherry picked some of them.
> 
> For context, pedantic has some... really pedantic lints like:

Yeah, I agree to skip it.

Comment (stefano-garzarella) [reply staging/vhost-device-sound/src/lib.rs:31]: > > Is it possible to enable these in some way for the entire workspace?
> 
> Good question. It's gonna be possible in stable soon:
> 
> https://rust-lang.github.io/rfcs/3389-manifest-lint.html
> 
> Quoting the RFC:
> 
> ```
> Currently, you can configure lints through
> 
>     #[<level>(<lint>)] or #![<level>(<lint>)], like #[forbid(unsafe)]
>         But this doesn't scale up with additional targets (benches, examples, tests) or workspaces
>     On the command line, like cargo clippy -- --forbid unsafe
>         This puts the burden on the caller
>     Through RUSTFLAGS, like RUSTFLAGS=--forbid=unsafe cargo clippy
>         This puts the burden on the caller
>     In .cargo/config.toml's target.*.rustflags
>         This couples you to the running in specific directories and not running in the right directory causes rebuilds
>         The cargo team has previously stated that [they would like to see package-specific config moved to manifests](https://internals.rust-lang.org/t/proposal-move-some-cargo-config-settings-to-cargo-toml/13336/14?u=epage)
> ```
> 
> So to enable them for the workspace we could put them in RUSTFLAGS for all targets in a `.cargo/config.toml`.

Cool, can you open a PR with these details, so we don't forget to do it when it's in stable?

File: staging/vhost-device-sound/Cargo.toml (hunk 1 above)
Comment (stefano-garzarella) [reply staging/vhost-device-sound/Cargo.toml:9]: Let's fix in another PR ;-)

File: staging/vhost-device-sound/src/lib.rs (hunk 2 above)
Comment (epilys) [reply staging/vhost-device-sound/src/lib.rs:31]: > Cool, can you open a PR with these details, so we don't forget to do it when it's in stable?

An issue you mean?

File: staging/vhost-device-sound/Cargo.toml (hunk 1 above)
Comment (epilys) [reply staging/vhost-device-sound/Cargo.toml:9]: #517

File: staging/vhost-device-sound/src/lib.rs (hunk 2 above)
Comment (stefano-garzarella) [reply staging/vhost-device-sound/src/lib.rs:31]: > > Cool, can you open a PR with these details, so we don't forget to do it when it's in stable?
> 
> An issue you mean?

Yep, sorry! Yesterday was a public holiday, so today is like Monday for me ;-P

**Pull request merged**